docs/proposals/device-crd.md

```diff
@@ -0,0 +1,457 @@
+    * [Device Actions](#device-actions)
+    * [Device Events](#device-events)
+   * [Security](#security)
+   * [Open questions](#open-questions)
+
+## Motivation
+
+Device management is a key feature required for IoT use-cases in edge computing.
+This proposal addresses how can we manage devices from the cloud and synchronize
+the device updates between edge nodes and cloud.
+
+### Goals
+
+Device management must :
+* provide APIs for managing devices from the cloud.
+* synchronize the device updates between cloud and edge nodes.
+
+### Non-goals
+
+* To design **secure** device provisioning.
+* To address OTA device firmware upgrades.
+* To address how device auto-discovery can happen.
+* To address device migration scenarios.
+
+## Proposal
+We propose using Kubernetes [Custom Resource Definitions (CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) to describe device metadata/status and a controller to synchronize these device updates between edge and cloud.
+<img src="../images/device-crd/device-crd-model.png">
+
+### Use Cases
+
+* Describe device properties.
+  * Users can describe device proeprties and access mechanisms to interact with / control the device.
+* Perform CRUD operations on devices from cloud.
+   * Users can create , update , delete device metadata from the cloud via the CRD APIs exposed by the Kubernetes API server.
+   * Users can control the expected state of a device via the CRD APIs.
+* Report device status from the edge node.
+  * Mapper applications running on the edge can report the device state as well the actual values of the control attributes.
+
+## Design Details
+
+### Device model CRD
+<img src="../images/device-crd/device-model-crd.png">
+
+A `device model` describes the device properties exposed by the device and property visitors to access these properties. A device model is like a reusable template using which many devices can be created and managed.
+
+The `Spec` contains the following fields
+
+| Field                 | Description                                                                      |
+|-----------------------|----------------------------------------------------------------------------------|
+| Properties            | A list of device properties with details like data type , access mode, units etc.|
+| Property Visitors     | A list of property visitors. Each visitor describes how to access a device property. e.g. to access the temperature of a CC2650 SensorTag , the uuid of the register to access is specified.|
+
+### Device model sample
+```yaml
+apiVersion: devices.kubeedge.io/v1alpha1
+kind: DeviceModel
+metadata:
+  labels:
+    description: 'TI Simplelink SensorTag Device Attributes Model'
+    manufacturer: 'Texas Instruments'
+    model: CC2650
+  name: sensor-tag-model
+spec:
+  properties:
+  - name: temperature
+    description: 'Temperature from Temperature Sensor in Degrees Fahrenheit'
+    metadata:
```

Comment:
I am not sure if we really need the metadata field. Theoretically, `name` can also be called metadata. Or can we come up with a better name?

Reply:
The property has a name which can be referenced by the property visitor to define additional access mechanisms. The metadata is a generic key -value map which can contain additional stuff like what is the data type , what is the min / max range , what is the units etc. Since we could have different types of devices with varying properties , this metadata can contain any of such information.